core/src/main/scala/org/ensime/indexer/domain.scala

```diff
@@ -35,46 +35,73 @@ final case class PackageName(path: List[String]) extends FullyQualifiedName {
   def parent = PackageName(path.init)
 }
 
-final case class ClassName(pack: PackageName, name: String) extends FullyQualifiedName with DescriptorType {
+sealed abstract class ClassName extends FullyQualifiedName with DescriptorType {
+  def pack: PackageName
+  def name: String
+
   def contains(o: FullyQualifiedName) = o match {
     case ClassName(op, on) if pack == op && on.startsWith(name) =>
       (on == name) || on.startsWith(name + "$")
     case MemberName(cn, _) => contains(cn)
     case _ => false
   }
-
-  def fqnString =
-    if (pack.path.isEmpty) name
-    else ClassName.cleanupPackage(pack.fqnString + "." + name)
-
-  def internalString = name match {
-    case "boolean" => "Z"
-    case "byte" => "B"
-    case "char" => "C"
-    case "short" => "S"
-    case "int" => "I"
-    case "long" => "J"
-    case "float" => "F"
-    case "double" => "D"
-    case "void" => "V"
-    case _ => "L" + (if (pack.path.isEmpty) name else pack.path.mkString("/") + "/" + name) + ";"
-  }
 }
 
 object ClassName {
+  def apply(pack: PackageName, name: String): ClassName =
+    if (pack.path.isEmpty && primitiveMapping.isDefinedAt(name))
+      Primitive(name, primitiveMapping(name))
+    else
+      NonPrimitive(pack, name)
+
+  def unapply(c: ClassName): Option[(PackageName, String)] = Some((c.pack, c.name))
+
+  private final case class NonPrimitive(pack: PackageName, name: String) extends ClassName {
```

Comment:
you could make these top level and make the constructor private to force people to go via the `ClassName` (trait companion).